mungegithub: Make approval label mimick lgtm label

apelisse opened the pull request:
Our goal is to move from **LGTM** to **approved** as the gating label for the submit-queue.
In order to transition smoothly, we will first begin by having both LGTM and Approved
labels that behave identically.  Initially, only LGTM will be needed for merge, though the approved label will be present, and eventually only the approved label will be needed.  

Notes: this implementation always adds and removes the two labels synchronously

The plan is to start warning about LGTM use later, indicating that its use will be phased out in favor of approved.  Then, it will be downgraded to be the prerequisite for approved.  

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/contrib/1889)

<!-- Reviewable:end -->

Commit:
mungegithub: Make approval label mimick lgtm label

Because we want to move from LGTM to approval to gate the submit-queue,
one of the preliminary step is to have LGTM and Approval label.
The plan is to start warning about LGTM use later, and then downgrade it
to just "first-step review".

foxish commented:
I'm concerned that a new munger which deals with `lgtm` introduces potential new races. Are we confident that this won't introduce any?

calebamiles commented:
that's at least something to think about, @foxish. maybe I take a look this evening or tomorrow morning (sorry my meeting chain continued unabated today) about adding in changes to the existing `lgtm*` mungers. How does that sound @apelisse?

apelisse commented:
@foxish, let's think about it. Here's what it does (or should do):
- Remove `lgtm` if `approved` was removed
- Remove `approved` if `lgtm` was removed
- Add `lgtm` if `approved` was added
- Add `approved` if `lgtm` was added

Today `approved` label is not used in any munger so no process can automatically remove or add it. The bot only add/remove the `lgtm` label so I don't see how we could end-up in a live-lock (but I agree that we must be careful how we proceed)

foxish commented on mungegithub/mungers/approve-lgtm.go:
```diff
@@ -0,0 +1,89 @@
+
+	if *lastLabel.Event == "labeled" {
+		obj.AddLabel("approved")
+		obj.AddLabel("lgtm")
```
Maybe I misunderstand this. If the last event was `labeled`, and the bot added the `approved` and `lgtm` labels, we would come here and find it and try to add the label again?

foxish replied:
As we discussed, [this is not a NOP](https://github.com/kubernetes/contrib/blob/f07b936a34401c5fb1af98573c69a0a392598639/mungegithub/github/github.go#L720-L750).

apelisse replied:
Should be addressed with the new commit

eparis replied:
I think every munger previously has does the `if !obj.HasLabel; obj.AddLabel`, but I have no problem changing the github.go code like you did to handle it in there.

It may be ever so slightly faster to make a slice and use `AddLabels()`, I know we'll only hit the api one time, but we'll do less label->set conversions and stuff...    That's just bikeshedding.

Also, no hard coded label strings allowed in mungers. Use lgtmLabel (and make one for approvedLabel)

foxish commented:
@apelisse I agree, there is no immediate race that could arise here. LGTM after the above comment is addressed.

Commit:
mungegithub: Do nothing if we add/remove existing/missing label

Currently, we will make a request to Github if we try to add an existing
label or to remove an unknown label. Let's verify that before making the
request.

Also move the count toward github api for AddLabels as we may return
early.

apelisse commented:
Hi @eparis, can you please verify that the "Do nothing if we add/remove existing/missing label" commit is OK?

eparis commented on mungegithub/github/github.go:
```diff
@@ -726,12 +726,17 @@ func (obj *MungeObject) AddLabel(label string) error {
 		return nil
 	}
+	if obj.LabelSet().IsSuperset(sets.NewString(labels...)) {
+		glog.Info("Labels are already set: quitting")
```
This seems too verbose for something common, no?

eparis commented on mungegithub/github/github.go:
```diff
@@ -754,6 +759,11 @@ func (obj *MungeObject) RemoveLabel(label string) error {
 	prNum := *obj.Issue.Number
 
+	if !obj.LabelSet().Has(label) {
+		glog.Info("Removing label that is not set: quitting")
```
Also this.

eparis commented on mungegithub/github/github.go:
```diff
@@ -726,12 +726,17 @@ func (obj *MungeObject) AddLabel(label string) error {
-	config.analytics.AddLabels.Call(config, nil)
 	glog.Infof("Adding labels %v to PR %d", labels, prNum)
 	if len(labels) == 0 {
 		glog.Info("No labels to add: quitting")
```
Not your fault, but you are here. This is a pretty bad error message. At least say the obj.Number and maybe the obj.Issue.Labels...

eparis commented on mungegithub/github/github.go:
```diff
@@ -726,12 +726,17 @@ func (obj *MungeObject) AddLabel(label string) error {
 	config := obj.config
 	prNum := *obj.Issue.Number
-	config.analytics.AddLabels.Call(config, nil)
 	glog.Infof("Adding labels %v to PR %d", labels, prNum)
```
This needs to be after the `IsSuperset()` early return (like you already did with the analytics collection)

eparis commented on mungegithub/mungers/approve-lgtm.go:
```diff
@@ -0,0 +1,89 @@
+		obj.RemoveLabel("approved")
+		obj.RemoveLabel("lgtm")
+	} else {
+		glog.Fatal("Received unexpected event: ", *lastLabel.Event)
```
This will crash the entire mungegithub.  Is that necessary? Seems like a `glog.Errorf()` and return would be sufficient instead of constantly crashing if github api does something unexpected.

apelisse replied:
Yes, we shouldn't crash if we receive an unexpected answer from github (I tend to like assertion-crash for programming errors though).

In this instance, you are correct, it will crash in the following scenario (amongst others): If github returns lgtm as a label, and event is neither label nor unlabeled, then it would crash.
I'll fix it.

apelisse commented:
@eparis I think I've addressed your comments. PTAL

apelisse commented:
We have decided to not do that anymore.

Closing.
